src/lib.rs

```
@@ -0,0 +1,349 @@
+impl PartialOrd for HeapItem {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for HeapItem {
+    fn eq(&self, other: &Self) -> bool {
+        self.event.time == other.event.time
+    }
+}
+
+impl Eq for HeapItem {}
+
+pub struct TraceReader {
+    heap: BinaryHeap<HeapItem>,
+    genesis: u64,
+}
+
+impl TraceReader {
+    pub fn new(device_path: &str) -> Result<Self, io::Error> {
+        let path = Path::new(device_path);
+        let dir = path.parent().ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "Invalid device path"))?;
+        let basename = path.file_name().ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "Invalid device path"))?.to_str().unwrap();
+        let prefix = format!("{}.blktrace.", basename);
+
+        let files: Vec<_> = fs::read_dir(dir)?
+            .filter_map(Result::ok)
+            .map(|e| e.path())
+            .filter(|p| p.is_file())
+            .filter(|p| {
+                p.file_name()
+                    .and_then(|s| s.to_str())
+                    .map_or(false, |s| s.starts_with(&prefix))
+            })
+            .collect();
+
+        if files.is_empty() {
+            return Err(io::Error::new(io::ErrorKind::NotFound, "No trace files found"));
+        }
+
+        let items: Result<Vec<Option<HeapItem>>, io::Error> = files.into_iter()
+            .map(|p| {
+                let mut trace_file_iterator = TraceFile::new(&p)?;
+                if let Some(event_result) = trace_file_iterator.next() {
+                    let event = event_result?;
+                    Ok(Some(HeapItem { event, trace_file_iterator }))
+                } else {
+                    Ok(None)
+                }
+            })
+            .collect();
+
+        let items: Vec<HeapItem> = items?.into_iter().filter_map(|x| x).collect();
+        let genesis = items.iter().min_by_key(|item| item.event.time).map_or(0, |item| item.event.time);
```

We probably want to unwrap here. If there are no items, we should return an error saying so before we construct the reader.